target/riscv/insn_trans/trans_cheri.inc.c

```diff
@@ -425,35 +425,75 @@ static inline bool trans_sc(DisasContext *ctx, arg_sc *a)
 
 
 // Atomic ops
-static inline bool trans_lr_cap(DisasContext *ctx, arg_lr_cap *a)
+static inline bool trans_lr_cap_impl(DisasContext *ctx, arg_atomic *a,
+                                     cheri_cap_cap_helper *helper)
 {
-    /* In a parallel context, stop the world and single step.  */
     if (tb_cflags(ctx->base.tb) & CF_PARALLEL) {
+        // In a parallel context, stop the world and single step.
         gen_helper_exit_atomic(cpu_env);
         ctx->base.is_jmp = DISAS_NORETURN;
     } else {
-        // Note: we ignore the Acquire/release flags since this using
+        // Note: we ignore the Acquire/release flags since using
         // helper_exit_atomic forces exlusive execution so we get SC semantics.
         tcg_debug_assert(a->rs2 == 0);
-        gen_cheri_cap_cap(a->rd, a->rs1, &gen_helper_lr_cap);
+        gen_cheri_cap_cap(a->rd, a->rs1, helper);
     }
     return true;
 }
 
-static inline bool trans_sc_cap(DisasContext *ctx, arg_sc_cap *a)
+static inline bool trans_lr_c(DisasContext *ctx, arg_lr_c *a)
 {
-    /* In a parallel context, stop the world and single step.  */
+    // Note: The capmode dependent address interpretation happens in the
```

Comment:
Do you think this is something you will be changing, to match the non-cap-width atomics?

Reply:
I'm not sure. It might make sense to do so but I don't it reduces code duplication very much.
Doing it all with cmpxchg will be difficult so we might as well do it all in the helper with an EXCP_ATOMIC